Config dev

Jake-Janssen opened the pull request.

Commit:
Config revisions, storing data, adjusting weight and saving locations

Commit:
allowing config.json for weights

Commit:
prep for open source

mpowelson commented on pcs_detection/scripts/data/aluminum_3/training_labels.xml:
```diff
@@ -0,0 +1,181 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Can we rename this dataset to example_dataset_1 or something similar. That's why I had those dataset directories. Also, could we split them into 2 separate datasets to show how to combine multiple?

mpowelson replied:
It doesn't have to be anything special. You could just copy the second half into the other directory

mpowelson commented on pcs_detection/scripts/data/fcn8_train.py:
```diff
@@ -0,0 +1,50 @@
+
+import json
+import os        
+from src_python.pcs_detection.process import test_dataloader, train
```
I realize you are not installing this with cmake, but when you do it will install the pcs_detection module without you having to do this. In fact, I'm a bit surprised this works unless you are doing something tricky with your Python path.

Long story, short please change `src_python.pcs_detection.process` to `pcs_detection.process`

Jake-Janssen replied:
Ah that makes sense, I could not get the import from pcs_detection.process to work so I ran it from the outer pcs_detection directory

mpowelson commented on pcs_detection/scripts/data/fcn8_validate.py:
```diff
@@ -0,0 +1,47 @@
+if __name__ == '__main__':
+    # Import Config json file and convert into format we need
+    dir_path = os.path.dirname(os.path.realpath(__file__))
+    with open(dir_path + '/weights/using_negatives_fcn8_GREY_19_10_10_101037/config.json') as json_data_file:
```
Can we rename this to something more reasonable? Like `example_weights` or `example_weld_detector_weights`?

Jake-Janssen replied:
renamed to example_weights

mpowelson commented on pcs_detection/scripts/fcn8_inference.py:
```diff
@@ -34,7 +34,7 @@
 
 # Inference module
-from pcs_detection.inference import Inference
+from src_python.pcs_detection.inference import Inference
```
See above

mpowelson commented on pcs_detection/src_python/pcs_detection/utils.py:
```diff
@@ -42,7 +42,7 @@ def get_labels_from_xml(label_path):
     # name of the outer directory this label corresponds to
     # allows for images in different directories to have the same filename 
-    dataset_name = label_path.split('/')[4]
+    dataset_name = label_path.split('/')[-2]
```
Is -2 always correct?

Jake-Janssen replied:
Should be with the current convention. I typically like to leave labels.xml in a directory above the actual images

mpowelson replied:
If that is a requirement we should put that in the readme.

Jake-Janssen replied:
This will be a requirement, which readme should it go in? The general one or the one specific to pcs_detection

mpowelson replied:
pcs_detection

mpowelson commented on pcs_detection/scripts/data/weights/.gitignore:
```diff
@@ -1,4 +1,2 @@
-* 
```
I believe this is needed to keep it from saving the weight file is it not?

mpowelson commented on pcs_detection/scripts/data/weights/using_negatives_fcn8_GREY_19_10_10_101037/config.json:
```diff
@@ -0,0 +1,60 @@
+    ],
+    "USE_FULL_IMAGE": false,
+    "VAL_WEIGHT_PATH": "38.h5",
+    "TRAINING_DIRS": [
```
Config is for infererence isn't it? Then it shouldn't need the training data, validation_data, learning rate, or background_reduction.

Jake-Janssen replied:
That's true, none of those parameters are needed to run. I typically save off everything the weights were trained on as a reference for future training.

mpowelson replied:
Yeah. Maybe you could dump that into a log file or something. Or even store it in the h5 file. It just seems a bit confusing what are actual inputs.

Jake-Janssen replied:
That makes sense, I'll put it in a log file

Jake-Janssen replied:
Two files are now saved off after training. One is the full training config which is used for the example_weights.py since data_loader.py requires most parameters for label reconstruction. I also made an inference_config that has only requirements for inference

mpowelson commented on pcs_detection/scripts/data/training_config.json:
```diff
@@ -0,0 +1,69 @@
+{
+    "MODE": "TRAIN",
+    "WEIGHT_ID": "new_repo_test",
+    "WEIGHT_DIR": "data/weights",
```
I would think this directory is relative to the location of the config? That doesn't look correct in this case. /scripts/data/data/weights isn't correct.

Jake-Janssen replied:
training config and script moved to scripts directory

mpowelson commented on pcs_detection/scripts/data/fcn8_validate.py:
```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env python3
+'''
+ * @file fcn8_validate.py
```
These should be in scripts not data.

mpowelson commented on pcs_detection/scripts/data/training_config.json:
```diff
@@ -0,0 +1,69 @@
+
+    "MODEL": "fcn8",
+
+    "CHANNEL": "GREY",
```
Could you create a readme in pcs_detection that describes these configs?

Jake-Janssen replied:
Added readme with descriptions of each config and an explanation for all config keys

mpowelson commented on pcs_detection/scripts/data/training_config.json:
```diff
@@ -0,0 +1,69 @@
+
+    "TRAINING_DIRS": [
+        {
+            "dir_path": "data/aluminum_3/training",
```
These paths are relative to what?

Jake-Janssen replied:
This is relative to being in the outer pcs_detection directory and running python3 scripts/data/fcn8_train.py
When using cmake are you able to run it from the data directory?

mpowelson replied:
I would say these paths need to be relative to the config file path. I have not tried to run it, but I would expect the scripts to go in /scripts not /scripts/data. Then the code should run regardless of where you are running it from.

Jake-Janssen replied:
config moved up a directory

mpowelson reviewed (changes requested):
The main thing that needs to be fixes is the imports. You may need to add pcs_detection to your path manually to get it to work, but a a person building this in cmake would have it installed at compile time.

Also, the configs don't look that cleaned up. I'd like to see the inference config be pretty simple. It shouldn't need paths to datasets or anything like that. All I'd imagine it needing is a path to the weights and info on what preprocessing to do.

Commit:
Renamed data directories
Modified import paths
all scripts moved to scripts
inference_config.json now saved off
more data added
slight tuning on confidence threshold
wrote config readme

Commit:
added requirements to the readme
